flow/util/genMetrics.py

```diff
@@ -68,8 +68,12 @@ def extractTagFromFile(jsonTag, jsonFile, pattern, file, count=False, occurrence
       content = f.read()
 
     m = re.findall(pattern, content, re.M)
+    useDefault = False
+    if occurrence != -1 and abs(occurrence) > len(m):
+      # Not that many occurrences, use default
+      useDefault = True
 
-    if m:
+    if m and not useDefault :
```

Comment:
What is the goal here?

Reply:
In case metrics is optional, may not always occur. In that case, when occurrence (when !=-1) is not found, this code leads to error. The newly added metrics in this PR are optional and not on-by-defalut. One of the metrics uses occurrence = -2. So when the metrics is not present, instead of using default, it errors out.

Reply:
This is not clear in the code. For the case of an optional metric my suggestion is to add a kwarg to extractTagFromFile, e.g., `required=True`, and use that to implement your logic.

Setting `occurrence = -2` says to the parser to get the second to last element, thus requiring that the parser finds at least two times the regex. It would be a good idea to add a test to avoid unwanted errors.

Reply:
So you mean to add another argument required=True? Check on occurrence would still be required since we do not know when required=False beforehand as it is dependent on settings in Makefile. Can we have dependency on Makefile variables here in genMetrics.py? I dont think would be right thing to do.

Reply:
> So you mean to add another argument required=True?

yes

> Check on occurrence would still be required since we do not know when required=False beforehand as it is dependent on settings in Makefile.

Does the code below solve the issue?

```
patternNotFound = (len(m) == 0)
if patternNotFound and not required:
     # handle here the case for optional metric not found
     # probably just a return to skip the metric (?)
```

>  Can we have dependency on Makefile variables here in genMetrics.py? I dont think would be right thing to do.

We could, be I agree that is not the best path.

Reply:
We are looking for second last occurrence (with occurrence = -2). So len(m) would not be zero. But len(m) < abs(occurence) could be true. "not required" check can be added, but would there be any else for that?

Reply:
> We are looking for second last occurrence (with occurrence = -2).

This works with the version on master already, so no new changes needed for this feature.

> So len(m) would not be zero. But len(m) < abs(occurence) could be true. 

The test you added to the `if` statement is the one that I think is not needed (and the source of my concern with code legibility).

> "not required" check can be added, but would there be any else for that?

I don't think an `else` would be needed. The `if` statement that checks if the metric is required should `return` when true or continue with current behavior when false.

Reply:
> > We are looking for second last occurrence (with occurrence = -2).
> 
> This works with the version on master already, so no new changes needed for this feature.
It works when 2 occurrences are there. If the file does not have 2 occurrences it errors. This is only reason of changing code here. We dont want to error out when optional metric is missing.
> 
> > So len(m) would not be zero. But len(m) < abs(occurence) could be true.
> 
> The test you added to the `if` statement is the one that I think is not needed (and the source of my concern with code legibility).
That test helps skip lookup resulting in error and lets code take default value in else part.
> 
> > "not required" check can be added, but would there be any else for that?
> 
> I don't think an `else` would be needed. The `if` statement that checks if the metric is required should `return` when true or continue with current behavior when false.

I think you have understood the intent behind code. Please suggest alternate code considering all these points.

Reply:
Add the argument:
```
def extractTagFromFile(jsonTag, jsonFile, pattern, file, count=False, occurrence=-1, defaultNotFound="N/A", t=str, required=True):
```

Replace the `if` statement you added in lines 71-74 with:
```
patternNotFound = (len(m) == 0)
if patternNotFound and not required:
    return
```

Revert the change you made to line 76
`if m and not useDefault :` -> `if m:`

Add `required=False` to the `extractTagFromFile(...)` calls as necessary.

Reply:
This approach can work, but need to check for len(m) < abs(occurrence) otherwise it would give error since m is 1 when call is made with required=False.
Also we are not adding default value in jsonFile, so number of rows in different variants (one with required=True and another required=False) would be different. Thus comparison would miss optional rows.

I'll make these 2 changes in addition and checkin.